jcl/src/openj9.dtfj/share/classes/com/ibm/dtfj/javacore/parser/j9/section/threadhistory/ThreadHistorySectionParser.java

```diff
@@ -0,0 +1,153 @@
+	 * @throws ParserException on failure to parse the history due to IOException or excessive line length
+	 */
+	@Override
+	protected void topLevelRule() throws ParserException {
+		fImageAddressSpaceBuilder = fImageBuilder.getCurrentAddressSpaceBuilder();
+		fImageProcessBuilder = fImageAddressSpaceBuilder.getCurrentImageProcessBuilder();
+		fJavaRuntimeBuilder = fImageProcessBuilder.getCurrentJavaRuntimeBuilder();
+
+		parseThreadHistoryLine();
+	}
+
+	/**
+	 * Parse the thread history information.
+	 * @throws ParserException on failure to parse the history due to IOException or excessive line length
+	 */
+	private void parseThreadHistoryLine() throws ParserException {
+		IAttributeValueMap results;
+
+		// Process the history lines
+		while ((results = processTagLineOptional(T_1XECTHTYPE)) != null) {
+			String name = results.getTokenValue(THREADHISTORY_NAME);
+			List<String> lines = new ArrayList<>();
+			boolean first = true;
+			IAttributeValueMap resultsTraceLine;
+			while ((resultsTraceLine = processTagLineOptional(T_3XEHSTTYPE)) != null) {
+				// 3XEHSTTYPE
+				// 04:36:21:235675300 GMT j9dmp.9 -   Preparing for dump, filename=C:\Users\andre\javacore.20220930.053619.81568.0001.txt
+				String hist = resultsTraceLine.getTokenValue(THREADHISTORY_LINE);
+				lines.add(hist);
+				if (first) {
+					first = false;
+					ImageBuilder imageBuilder = (ImageBuilder)fImageBuilder;
+					try {
+						long oldtime = imageBuilder.getImage().getCreationTime();
+						long newtime = TraceTimeParser.parseTimeAdjust(hist, oldtime, 15);
+						if (oldtime != newtime) {
+							fImageBuilder.setCreationTime(newtime);
+						}
+					} catch (DataUnavailable e) {
+						continue;
+					}
+				}
+			}
+			// lines appear in Javacore as newest first, so reverse
+			Collections.reverse(lines);
+			fJavaRuntimeBuilder.addTraceBuffer(name, lines);
+		}
+
+		/*
+		 * Fix up JVM start time.
+		 * The start time doesn't have a time zone, so is parsed according
+		 * to the parser's timezone, which might differ from the Javacore.
+		 * Once the javacore time is fixed up, the offset between the
+		 * JVM start time and nanos should be the same as the offset between
+		 * the javacore time and nanos.
+		 */
+		ImageBuilder imageBuilder = (ImageBuilder)fImageBuilder;
+		JavaRuntimeBuilder javaRuntimeBuilder = (JavaRuntimeBuilder)fJavaRuntimeBuilder;
+		try {
+			/*
+			 * The seconds and milliseconds are presumed correct, but due to timezone problems
+			 * the hours might be wrong, and minutes in steps of 15 minutes might be wrong.
+			 */
+			final long roundingMillis = 15 * 60 * 1000;
+			long creationTime = imageBuilder.getImage().getCreationTime();
+			long creationNanos = imageBuilder.getImage().getCreationTimeNanos();
+			long startTime = javaRuntimeBuilder.getStartTime();
+			long startNanos = javaRuntimeBuilder.getStartTimeNanos();
+			long javacoreDelta = creationTime - creationNanos / 1000000;
+			long startDelta = startTime - startNanos / 1000000;
```

Comment:
Missing parentheses on these two lines and also line 138.

Reply:
Parentheses added even though not necessary with operator precedence.